cmd/registry/cmd/compute-references.go

```diff
@@ -0,0 +1,144 @@
+	openapi_v3 "github.com/googleapis/gnostic/openapiv3"
+	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/proto"
+)
+
+func init() {
+	computeCmd.AddCommand(computeReferencesCmd)
+}
+
+var computeReferencesCmd = &cobra.Command{
+	Use:   "references",
+	Short: "Compute references of API specs",
+	Args:  cobra.MinimumNArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := context.TODO()
+		client, err := connection.NewClient(ctx)
+		if err != nil {
+			log.Fatalf("%s", err.Error())
+		}
+		// Initialize task queue.
+		taskQueue := make(chan core.Task, 1024)
+		workerCount := 64
+		for i := 0; i < workerCount; i++ {
+			core.WaitGroup().Add(1)
+			go core.Worker(ctx, taskQueue)
+		}
+		// Generate tasks.
+		name := args[0]
+		if m := names.SpecRegexp().FindStringSubmatch(name); m != nil {
+			// Iterate through a collection of specs and compute references for each
+			err = core.ListSpecs(ctx, client, m, computeFilter, func(spec *rpc.ApiSpec) {
+				taskQueue <- &computeReferencesTask{
+					ctx:      ctx,
+					client:   client,
+					specName: spec.Name,
+				}
+			})
+			if err != nil {
+				log.Fatalf("%s", err.Error())
+			}
+			close(taskQueue)
+			core.WaitGroup().Wait()
+		}
+	},
+}
+
+type computeReferencesTask struct {
+	ctx      context.Context
+	client   connection.Client
+	specName string
+}
+
+func (task *computeReferencesTask) Name() string {
+	return "compute references " + task.specName
+}
+
+func (task *computeReferencesTask) Run() error {
+	request := &rpc.GetApiSpecRequest{
+		Name: task.specName,
+		View: rpc.View_FULL,
+	}
+	spec, err := task.client.GetApiSpec(task.ctx, request)
+	if err != nil {
+		return err
+	}
+	relation := "references"
+	log.Printf("computing %s/properties/%s", spec.Name, relation)
+	var references *rpc.References
+	if core.IsOpenAPIv2(spec.GetMimeType()) {
+		data, err := core.GetBytesForSpec(spec)
+		if err != nil {
+			return nil
+		}
+		document, err := openapi_v2.ParseDocument(data)
+		if err != nil {
+			return fmt.Errorf("invalid OpenAPI: %s", spec.Name)
+		}
+		references, err = core.NewReferencesFromOpenAPIv2(document)
+		if err != nil {
+			return err
+		}
+	} else if core.IsOpenAPIv3(spec.GetMimeType()) {
+		data, err := core.GetBytesForSpec(spec)
+		if err != nil {
+			return nil
+		}
+		document, err := openapi_v3.ParseDocument(data)
+		if err != nil {
+			return fmt.Errorf("invalid OpenAPI: %s", spec.Name)
+		}
+		references, err = core.NewReferencesFromOpenAPIv3(document)
+		if err != nil {
+			return err
+		}
+	} else if core.IsProto(spec.GetMimeType()) && core.IsZipArchive(spec.GetMimeType()) {
+		references, err = core.NewReferencesFromZippedProtos(spec.GetContents())
+		if err != nil {
+			return fmt.Errorf("error processing protos: %s", spec.Name)
+		}
+	} else {
+		return fmt.Errorf("we don't know how to compute references for %s", spec.Name)
+	}
+	if references == nil {
+		return fmt.Errorf("we don't know how to compute references for %s", spec.Name)
+	}
```

Comment:
Do we intend to have `nil` be a valid non-error return value from NewReferencesFromX methods? This seems unusual to me. IMO those methods should be responsible for checking this and return an error when they cannot compute the references.

Reply:
I think that's a reasonable expectation - I'd say this was just defensive programming.